js/src/tooltip.js

```diff
@@ -35,7 +38,10 @@ const DefaultType = {
   offset            : '(number|string|function)',
   container         : '(string|element|boolean)',
   fallbackPlacement : '(string|array)',
-  boundary          : '(string|element)'
+  boundary          : '(string|element)',
+  sanitizeTemplate  : 'boolean',
```

Comment:
There's no line yet in the PR for the code I'm talking about, so I'll just comment it here.

Any option that has to deal with sanitization must be excluded from being able to be supplied by data attributes.

Otherwise, it defeats the whole point of a sanitization method because they could simply disable it entirely or supply their own whitelist (e.g. `data-sanitize="false"`).

Currently, it just merges them in (https://github.com/twbs/bootstrap/blob/v4-dev/js/src/tooltip.js#L641):
```js
  _getConfig(config) {
    config = {
      ...this.constructor.Default,
      ...$(this.element).data(),
      ...typeof config === 'object' && config ? config : {}
    }
    // ...
  }
``` 

These needs to be a way to exclude specific data attributes so they can only be supplied via default values or the JavaScript initiator:

```js
  _getConfig(config, excludeData = []) {
    const data = $(this.element).data()
    for (let i = 0, l = excludeData.length; i < l; i++) {
      delete data[excludeData[i]]
    }
    config = {
      ...this.constructor.Default,
      ...data,
      ...typeof config === 'object' && config ? config : {}
    }
    // ...
  }
```

Reply:
Unfortunately we cannot do that, Bootstrap works with data API.

People which use our plugins can force `sanitizeTemplate` if they need it

Reply:
Then this entire PR is pointless...

Reply:
Sanitize string on a front library is pointless too, but it can prevent things 👍

As you can see here: https://getbootstrap.com/docs/4.3/getting-started/javascript/#data-attributes

You can disable the data API for specific plugins or for all of our plugins.

Or you can force this value to always be `true` because options passed in JavaScript override those passed by HTML.

```js
$('#myTooltip').tooltip({
  sanitize template: true,
})
```
It will be always true even if it's at false on the HTML with data attribute.

Personally I think it would be a bit strange to disable that 😣

/CC @XhmikosR what do you think about that ?

Reply:
> Sanitize string on a front library is pointless too

No, it's not... it's a requirement if it creates content (like this is doing). That's why is called a vulnerability. Are you even taking this seriously?

> It will be always true even if it's at false on the HTML with data attribute.

Except that is not how secure code __should__ work. It should be __secure by default__. A user shouldn't have to "opt in" for protection.

The reality is that the way this framework handles data attributes are backward, to begin with. Almost any other JS plugin allows data attributes to override global and supplied options. That is because data attributes are a local element configuration.

Because of this backward way of handling data attributes, other code (like http://cgit.drupalcode.org/bootstrap/tree/js/bootstrap.js?h=7.x-3.x#n175) has to supply data attributes on top of their dynamic CMS settings just so they can work at all.

So yes, data attributes could still make their way through; especially if there existing code (e.g. a simple "update" wouldn't actually "secure" anything at all until that manually supplied said forced override... which is absolutely ridiculous).

> Personally I think it would be a bit strange to disable that 😣

Why? There is absolutely no reason sanitization should be configurable at an element level! This is about security and this kind of option should only ever be supplied by trusted sources. Data attributes are not "trusted sources".

Reply:
You cut my sentence but anyway...

We are taking that seriously, we built a solution in one day, that's not a complete solution but we work on it... So stay kind in your words and manners.

I think it's a bit strange because we allow that everywhere in Bootstrap and doing that for just one option feel a bit strange to me, but we can make exceptions.

Reply:
> We are taking that seriously, we built a solution in one day, that's not a complete solution but we work on it

I am certainly impressed, no doubt. But there is a certain level of responsibility once one decides to take on a topic such as this. I know this framework doesn't like to bloat the codebase.

It's partially why I originally suggested simply disabling the data attribute entirely unless a 3rd party library was present. That way this framework wouldn't be responsible for sanitization at all and could offload it to something else; while also keeping security in mind.

I really do appreciate the effort and responsiveness here.

> So stay kind in your words and manners.

I'm sorry. It just felt like you weren't reading my comments and quickly dismissing them. I spent a lot of time reviewing this and writing up suggestions, which have logical explanations and reasoning behind them. I'm very familiar with the FE and XSS related issues; this isn't my first rodeo either 😉 

> I think it's a bit strange because we allow that everywhere in Bootstrap and doing that for just one option feel a bit strange to me, but we can make exceptions.

Security reasons are always an exception.

Reply:
Even if I dismiss your comments I'm always open to speak about it. BTW some of them I reopened your comments 😉

If you present to me why you suggested that I'll think about it that's sure.
Sometimes when you suggest something to someone you have to add a bit of context why it should be like this.

I don't want to add a third library because folks which use Bootstrap with NPM will still needs to install the third library even if they don't need it. So I made the decision to made a simple sanitizer which can be disabled if folks now how to sanitize themselves their HTML

Reply:
> BTW some of them I reopened your comments 😉

I appreciate that



> Sometimes when you suggest something to someone you have to add a bit of context why it should be like this.

I have, that's why I wrote a lot.



> I don't want to add a third library

I'm not saying that it should be a requirement. I'm saying that it **could** just be supported. This isn't a real example, but it may be better (language wise) for me to simply give you a simplistic if/else block:

```js
// Support https://jsxss.com (or something else, or multiple, etc.).
if (config.html && typeof window.filterXSS === 'function') {
  // The `data-template`, `data-title` and `data-content`  attributes would
  // be allowed to use `$.html()` or `innerHTML`  by supplying sanitized
  // content using filterXSS().
}
else {
  // The `data-template`, `data-title` and `data-content`  attributes would
  // only be allowed to use `$.text()` or `innerText`  by supplying  unsanitized
  // content based on what was passed. This would essentually "break"
  // `data-template` until they supplied a supported XSS filter library.
}
```

Users could enable it, but it would only truely become enabled whether a supported 3rd party sanitization library is actually present.

Regardless, the LOC bump here is actually quite small. If y'all are happy with supporting a simplistic whitelisted built-in approach, that would most certainly be better. We can always replace it with something a little more hefty if we need.